azure-iot-hub/tests/test_iothub_registry_manager.py

```diff
@@ -0,0 +1,800 @@
+
+
+@pytest.fixture(scope="function")
+def iothub_registry_manager():
+    connection_string = "HostName={hostname};DeviceId={device_id};SharedAccessKeyName={skn};SharedAccessKey={sk}".format(
+        hostname=fake_hostname,
+        device_id=fake_device_id,
+        skn=fake_shared_access_key_name,
+        sk=fake_shared_access_key,
+    )
+    iothub_registry_manager = IoTHubRegistryManager(connection_string)
+    return iothub_registry_manager
+
+
+@pytest.mark.describe("IoTHubRegistryManager")
+class TestIoTHubRegistryManagerInit(object):
+    @pytest.mark.it("instantiates correctly with auth")
+    def test_init_auth(self, iothub_registry_manager):
+        assert iothub_registry_manager is not None
+        assert iothub_registry_manager.auth is not None
+        assert isinstance(iothub_registry_manager.auth, ConnectionStringAuthentication)
+
+    @pytest.mark.it("instantiates correctly with protocol client with correct Service API version")
+    def test_init_protocol_client(self, iothub_registry_manager):
+        assert iothub_registry_manager is not None
+        assert iothub_registry_manager.protocol is not None
+        assert isinstance(iothub_registry_manager.protocol, IotHubGatewayServiceAPIs20190630)
+
+
+different_devices_sas = [
+    pytest.param(
+        {
+            "function_name": "create_device_with_sas",
+            "init_kwargs": {
+                "device_id": fake_device_id,
+                "primary_key": fake_primary_key,
+                "secondary_key": None,
+                "status": fake_status,
+            },
+            "etag": None,
+        },
+        id="Create with Primary Key",
+    ),
+    pytest.param(
+        {
+            "function_name": "create_device_with_sas",
+            "init_kwargs": {
+                "device_id": fake_device_id,
+                "primary_key": None,
+                "secondary_key": fake_secondary_key,
+                "status": fake_status,
+            },
+            "etag": None,
+        },
+        id="Create with Secondary Key",
+    ),
+    pytest.param(
+        {
+            "function_name": "update_device_with_sas",
+            "init_kwargs": {
+                "device_id": fake_device_id,
+                "primary_key": fake_primary_key,
+                "secondary_key": None,
+                "status": fake_status,
+                "etag": fake_etag,
+            },
+            "etag": fake_etag,
+        },
+        id="Update with Primary Key",
+    ),
+    pytest.param(
+        {
+            "function_name": "update_device_with_sas",
+            "init_kwargs": {
+                "device_id": fake_device_id,
+                "primary_key": None,
+                "secondary_key": fake_secondary_key,
+                "status": fake_status,
+                "etag": fake_etag,
+            },
+            "etag": fake_etag,
+        },
+        id="Update with Secondary Key",
+    ),
+]
```

Comment:
Using structures like this to parametrize tests is an antipattern that we are trying to avoid now due to inconsistency in output and difficulty of long term maintenance.

Parametrization should only be used for inputs to a unit under test, not for structuring tests themselves.